### PR TITLE
sink/codec(cdc): refactor Avro codec

### PR DESCRIPTION
**Pull request by zhangyangyu**
### What problem does this PR solve?

Issue Number: ref #5338 

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by zhaoxinyu** [comment]
@zhangyangyu , there are some code conflicts need to be resolved.

File: cdc/model/sink.go [hunk 1]
```diff
@@ -330,6 +330,33 @@ func (r *RowChangedEvent) HandleKeyColumns() []*Column {
...
+
+	for i, col := range cols {
+		if col != nil && col.Flag.IsHandleKey() {
+			pkeyCols = append(pkeyCols, col)
+			pkeyColInfos = append(pkeyColInfos, r.ColInfos[i])
+		}
+	}
+
+	if len(pkeyCols) == 0 {
+		// TODO redact the message
```
Comment (hi-rustin) [inline cdc/model/sink.go:353]: Can we remove this todo?

File: cdc/sink/codec/avro.go [hunk 2]
```diff
@@ -19,88 +19,76 @@ import (
...
 // AppendRowChangedEvent appends a row change event to the encoder
 // NOTE: the encoder can only store one RowChangedEvent!
-func (a *AvroEventBatchEncoder) AppendRowChangedEvent(e *model.RowChangedEvent) error {
-	mqMessage := NewMQMessage(config.ProtocolAvro, nil, nil, e.CommitTs, model.MqMessageTypeRow, &e.Table.Schema, &e.Table.Table)
+func (a *AvroEventBatchEncoder) AppendRowChangedEvent(
+	ctx context.Context,
+	e *model.RowChangedEvent,
+	topic string,
+) error {
+	log.Debug("AppendRowChangedEvent", zap.Reflect("rowChangedEvent", e))
```
Comment (hi-rustin) [inline cdc/sink/codec/avro.go:64]: Can we use `zap.Any` here?

File: cdc/sink/codec/avro.go [hunk 3]
```diff
@@ -19,88 +19,76 @@ import (
...
+		&e.Table.Table,
+	)
+	topic = sanitizeTopic(topic)
 
 	if !e.IsDelete() {
-		res, err := avroEncode(e.Table, a.valueSchemaManager, e.TableInfoVersion, e.Columns, e.ColInfos, a.tz)
+		res, err := a.avroEncode(ctx, e, topic, false)
 		if err != nil {
-			log.Warn("AppendRowChangedEvent: avro encoding failed", zap.String("table", e.Table.String()))
+			log.Warn(
```
Comment (hi-rustin) [inline cdc/sink/codec/avro.go:79]: Do we need to use log.Error here?

File: cdc/sink/codec/avro.go [hunk 4]
```diff
@@ -160,30 +152,86 @@ func (a *AvroEventBatchEncoder) Size() int {
...
+		cols, colInfos = e.HandleKeyColInfos()
+		enableTiDBExtension = false
+		schemaManager = a.keySchemaManager
+	} else {
+		cols = e.Columns
+		colInfos = e.ColInfos
+		enableTiDBExtension = a.enableTiDBExtension
+		schemaManager = a.valueSchemaManager
+		if e.IsInsert() {
+			operation = "c"
```
Comment (hi-rustin) [inline cdc/sink/codec/avro.go:176]: These are best if they can be constants, and if that seems too overkill, then there should at least be a comment explaining exactly how this one letter is understood.

File: cdc/sink/codec/avro.go [hunk 5]
```diff
@@ -193,41 +241,155 @@ func avroEncode(table *model.TableName, manager *AvroSchemaManager, tableVersion
...
+				sanitizedName,
+			),
+		)
+	}
+	return sanitizedName
+}
+
+// escape ".", it may has special meanings for sink connectors
+func sanitizeTopic(name string) string {
+	return strings.ReplaceAll(name, ".", "_")
```
Comment (hi-rustin) [inline cdc/sink/codec/avro.go:336]: Can we use replacementChar here?

File: cdc/sink/codec/avro.go [hunk 6]
```diff
@@ -193,41 +241,155 @@ func avroEncode(table *model.TableName, manager *AvroSchemaManager, tableVersion
...
+	return option
+}
+
+func getAvroNamespace(namespace string, tableName *model.TableName) string {
+	return sanitizeName(namespace) + "." + sanitizeName(tableName.Schema)
+}
+
+type avroSchema struct {
+	Type       string            `json:"type"`
+	Parameters map[string]string `json:"connect.parameters"`
```
Comment (hi-rustin) [inline cdc/sink/codec/avro.go:355]: ``json:"connect.parameters"`` Is this a typo? Or do we have to write it that way?

File: cdc/sink/codec/avro.go [hunk 7]
```diff
@@ -238,233 +400,260 @@ func ColumnInfoToAvroSchema(name string, columnInfo []*model.Column) (string, er
...
+	tt := getTiDBTypeFromColumn(col)
 	switch col.Type {
-	case mysql.TypeFloat:
-		return "float", nil
-	case mysql.TypeDouble:
-		return "double", nil
-	case mysql.TypeVarchar, mysql.TypeString, mysql.TypeVarString:
-		if col.Flag.IsBinary() {
-			return "bytes", nil
+	case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24: // BOOL/TINYINT/SMALLINT/MEDIUMINT
```
Comment (hi-rustin) [inline cdc/sink/codec/avro.go:478]:
```suggestion
        // BOOL/TINYINT/SMALLINT/MEDIUMINT
	case mysql.TypeTiny, mysql.TypeShort, mysql.TypeInt24: 
```

File: cdc/sink/codec/avro.go [hunk 8]
```diff
@@ -238,233 +400,260 @@ func ColumnInfoToAvroSchema(name string, columnInfo []*model.Column) (string, er
...
-			return unsignedLongAvroType, nil
+		return avroSchema{
+			Type: "bytes",
+			Parameters: map[string]string{
+				tidbType: tt,
+				"length": strconv.Itoa(displayFlen),
+			},
+		}, nil
+	case mysql.TypeNewDecimal:
+		if decimalHandlingMode == "precise" {
```
Comment (hi-rustin) [inline cdc/sink/codec/avro.go:523]: `precise` Better if it's a constant?

File: cdc/sink/codec/avro.go [hunk 9]
```diff
@@ -238,233 +400,260 @@ func ColumnInfoToAvroSchema(name string, columnInfo []*model.Column) (string, er
...
+			defaultFlen, defaultDecimal := mysql.GetDefaultFieldLengthAndDecimal(ft.Tp)
+			displayFlen, displayDecimal := ft.Flen, ft.Decimal
+			if displayFlen == -1 {
```
Comment (hi-rustin) [inline cdc/sink/codec/avro.go:526]: `-1` confusing.

File: cdc/sink/codec/interface.go [hunk 10]
```diff
@@ -31,7 +31,7 @@ type EventBatchEncoder interface {
 	// This event will be broadcast to all partitions to signal a global checkpoint.
 	EncodeCheckpointEvent(ts uint64) (*MQMessage, error)
 	// AppendRowChangedEvent appends a row changed event into the batch
-	AppendRowChangedEvent(e *model.RowChangedEvent) error
+	AppendRowChangedEvent(context.Context, *model.RowChangedEvent, string) error
```
Comment (hi-rustin) [inline cdc/sink/codec/interface.go:34]: A note on the implementation and usage of ctx and topic would be nice here!

File: cdc/sink/codec/interface.go [hunk 11]
```diff
@@ -150,22 +165,25 @@ type EncoderBuilder interface {
...
+		return newAvroEventBatchEncoderBuilder(ctx, c)
 	case config.ProtocolMaxwell:
 		return newMaxwellEventBatchEncoderBuilder(), nil
 	case config.ProtocolCanalJSON:
 		return newCanalFlatEventBatchEncoderBuilder(c), nil
 	case config.ProtocolCraft:
 		return newCraftEventBatchEncoderBuilder(c), nil
 	default:
-		log.Warn("unknown codec protocol value of EventBatchEncoder, use open-protocol as the default", zap.Any("protocolValue", int(c.protocol)))
+		log.Warn(
```
Comment (hi-rustin) [inline cdc/sink/codec/interface.go:183]: This should not be possible to execute, can we return an error here indicating that the type is not known?

File: cdc/sink/mq/dispatcher/event_router.go [hunk 12]
```diff
@@ -249,17 +250,32 @@ func getPartitionDispatcher(
...
 
 	// check if this rule is a valid topic expression
 	topicExpr := topic.Expression(ruleConfig.TopicRule)
-	err := topicExpr.Validate()
-	if err != nil {
-		return nil, err
+
+	if protocol != "" {
+		var p config.Protocol
+		if err := p.FromString(protocol); err != nil {
```
Comment (hi-rustin) [inline cdc/sink/mq/dispatcher/event_router.go:264]: It should have been done in the configuration check?

File: cdc/sink/mq/dispatcher/topic/expression.go [hunk 13]
```diff
@@ -30,6 +30,10 @@ var (
 	schemaRE = regexp.MustCompile(`\{schema\}`)
 	// tableRE is used to match substring '{table}' in topic expression
 	tableRE = regexp.MustCompile(`\{table\}`)
+	// avro has different topic name pattern requirements
```
Comment (hi-rustin) [inline cdc/sink/mq/dispatcher/topic/expression.go:33]: Please provide more detailed comments on it, thank you! For example, the difference from the original?

File: go.mod [hunk 14]
```diff
@@ -95,6 +95,8 @@ require (
 	upper.io/db.v3 v3.7.1+incompatible
 )
 
+require github.com/pkg/errors v0.9.1 // indirect
```
Comment (hi-rustin) [inline go.mod:98]: Why we need change this?

File: cdc/model/sink.go (hunk 1 above)
Comment (zhangyangyu) [reply cdc/model/sink.go:353]: copied from `HandleKeyColumns`. Not sure what to do.

File: cdc/model/sink.go [hunk 15]
```diff
@@ -330,6 +330,33 @@ func (r *RowChangedEvent) HandleKeyColumns() []*Column {
...
+		log.Panic("Cannot find handle key columns, bug?", zap.Reflect("event", r))
```
Comment (3AceShowHand) [inline cdc/model/sink.go:354]: `zap.Any` is always a better choice than `zap.Relect`.

File: cdc/sink/codec/avro.go [hunk 16]
```diff
@@ -160,30 +152,86 @@ func (a *AvroEventBatchEncoder) Size() int {
...
+	ctx context.Context,
+	e *model.RowChangedEvent,
+	topic string,
+	key bool,
+) (*avroEncodeResult, error) {
+	var cols []*model.Column
+	var colInfos []rowcodec.ColInfo
+	var enableTiDBExtension bool
+	var schemaManager *AvroSchemaManager
+	var operation string
```
Comment (3AceShowHand) [inline cdc/sink/codec/avro.go:165]: they can be grouped together.

File: cdc/sink/codec/avro.go [hunk 17]
```diff
@@ -160,30 +152,86 @@ func (a *AvroEventBatchEncoder) Size() int {
...
+			e.Table.Table,
+			cols,
+			colInfos,
+			enableTiDBExtension,
+			a.decimalHandlingMode,
+			a.bigintUnsignedHandlingMode,
+		)
 		if err != nil {
 			return "", errors.Annotate(err, "AvroEventBatchEncoder: generating schema failed")
 		}
```
Comment (3AceShowHand) [inline cdc/sink/codec/avro.go:199]: the error here, can only be `cerror.WrapError(cerror.ErrAvroMarshalFailed, err)`.

just use `errors.Trace(err)` should be ok.

File: cdc/sink/codec/schema_registry.go [hunk 18]
```diff
@@ -292,13 +372,23 @@ func (m *AvroSchemaManager) GetCachedOrRegister(ctx context.Context, tableName m
...
+		)
 	}
 
-	id, err := m.Register(ctx, tableName, codec)
+	id, err := m.Register(ctx, topicName, codec)
 	if err != nil {
 		return nil, 0, errors.Annotate(
-			cerror.WrapError(cerror.ErrAvroSchemaAPIError, err), "GetCachedOrRegister: Could not register schema")
+			cerror.WrapError(
+				cerror.ErrAvroSchemaAPIError,
```
Comment (3AceShowHand) [inline cdc/sink/codec/schema_registry.go:387]: ditto

File: cdc/sink/codec/interface.go (hunk 11 above)
Comment (zhangyangyu) [reply cdc/sink/codec/interface.go:183]: This would break the current logic using open-protocol as the default protocol. I think it's not relevant to avro and better leave it not changed now.

File: cdc/sink/codec/avro.go [hunk 19]
```diff
@@ -160,30 +152,86 @@ func (a *AvroEventBatchEncoder) Size() int {
...
 	if err != nil {
 		return nil, errors.Annotate(err, "AvroEventBatchEncoder: converting to native failed")
 	}
 
 	bin, err := avroCodec.BinaryFromNative(nil, native)
 	if err != nil {
 		return nil, errors.Annotate(
-			cerror.WrapError(cerror.ErrAvroEncodeToBinary, err), "AvroEventBatchEncoder: converting to Avro binary failed")
+			cerror.WrapError(
+				cerror.ErrAvroEncodeToBinary,
```
Comment (3AceShowHand) [inline cdc/sink/codec/avro.go:230]: `errors.Annotate` is verbose here, just `errors.Trace(err)` should be ok. add a log to help debug the problem if necessary.

File: cdc/sink/codec/avro.go (hunk 9 above)
Comment (zhangyangyu) [reply cdc/sink/codec/avro.go:526]: :-( that's what TiDB returns, not much we could do. I'd like readers to learn the complex TiDB default handling logic.

File: cdc/sink/mq/dispatcher/event_router.go (hunk 12 above)
Comment (zhangyangyu) [reply cdc/sink/mq/dispatcher/event_router.go:264]: That's the sad part. The protocol is parsed and simply used to new encoder config, not stored inside replica config. So at the time of newMqSink, the parsed protocol is not available. I don't understand why actually.

Here, I re-parse the protocol to ultilize the parsing logic in case anything might change.

File: go.mod (hunk 14 above)
Comment (zhangyangyu) [reply go.mod:98]: I changed testify to 1.7.1 and it appears automaticllay.

File: pkg/cmd/cli/cli_changefeed_create.go [hunk 20]
```diff
@@ -268,6 +271,7 @@ func (o *createChangefeedOptions) completeCfg(
 			// TODO(neil) enable ID bucket.
 		}
 	}
+	cfg.SchemaRegistry = o.schemaRegistry
```
[the remark this answers is not in the document]
Comment (zhangyangyu) [reply pkg/cmd/cli/cli_changefeed_create.go:274]: it will be checked in `newschemaregistry`, not only format but also connectivity.

File: pkg/config/mq_sink_protocol.go [hunk 21]
```diff
@@ -47,6 +47,8 @@ func (p *Protocol) FromString(protocol string) error {
 		*p = ProtocolCanal
 	case "avro":
 		*p = ProtocolAvro
+	case "flat-avro":
```
[the remark this answers is not in the document]
Comment (zhangyangyu) [reply pkg/config/mq_sink_protocol.go:50]: Seems I miss a UT here?

File: cdc/sink/codec/avro.go [hunk 22]
```diff
@@ -193,41 +245,156 @@ func avroEncode(table *model.TableName, manager *AvroSchemaManager, tableVersion
...
+				"Name '%s' potentially not safe for serialization, replaced with '%s'",
+				name,
+				sanitizedName,
+			),
+		)
+	}
+	return sanitizedName
+}
+
+// escape ".", it may has special meanings for sink connectors
```
[the remark this answers is not in the document]
Comment (zhangyangyu) [reply cdc/sink/codec/avro.go:338]: rather stupid especially this is a must for public functions. Leading to many comments like "// WithMaxMessageBytes set the `maxMessageBytes`", "// Validate the Config".

File: cdc/sink/codec/avro.go [hunk 23]
```diff
@@ -193,41 +233,157 @@ func avroEncode(table *model.TableName, manager *AvroSchemaManager, tableVersion
...
+			changed = true
+		} else {
+			sb.WriteRune(c)
+		}
+	}
+
+	sanitizedName := sb.String()
+	if changed {
+		log.Warn(
+			fmt.Sprintf(
```
Comment (3AceShowHand) [inline cdc/sink/codec/avro.go:317]: this log message can be replaced by `Name is potentially not safe for serialization, replace it, zap.String("name", name), zap.String("replacedName", sanitizedName).

File: cdc/sink/codec/config_test.go [hunk 24]
```diff
@@ -17,127 +17,191 @@ import (
...
+	require.Equal(t, "", c.avroSchemaRegistry)
 }
 
 func TestConfigApplyValidate(t *testing.T) {
 	t.Parallel()
+
+	// enable-tidb-extension
 	uri := "kafka://127.0.0.1:9092/abc?protocol=canal-json&enable-tidb-extension=true"
 	sinkURI, err := url.Parse(uri)
-	require.Nil(t, err)
```
Comment (3AceShowHand) [inline cdc/sink/codec/config_test.go:40]: what is the difference between `Nil` and `NoError` here, can we revert this change ?

File: cdc/sink/codec/schema_registry.go [hunk 25]
```diff
@@ -71,68 +69,102 @@ type lookupResponse struct {
...
 	registryURL = strings.TrimRight(registryURL, "/")
-	// Test connectivity to the Schema Registry
 	httpCli, err := httputil.NewClient(credential)
 	if err != nil {
 		return nil, errors.Trace(err)
 	}
 	resp, err := httpCli.Get(ctx, registryURL)
 	if err != nil {
 		return nil, errors.Annotate(
-			cerror.WrapError(cerror.ErrAvroSchemaAPIError, err), "Test connection to Schema Registry failed")
```
Comment (3AceShowHand) [inline cdc/sink/codec/schema_registry.go:87]: can you also help remove `errors.Annotate` here ?

File: cdc/sink/codec/schema_registry.go [hunk 26]
```diff
@@ -71,68 +69,102 @@ type lookupResponse struct {
...
+			),
+			"Test connection to Schema Registry failed",
+		)
 	}
 	defer resp.Body.Close()
 
 	text, err := io.ReadAll(resp.Body)
 	if err != nil {
 		return nil, errors.Annotate(
-			cerror.WrapError(cerror.ErrAvroSchemaAPIError, err), "Reading response from Schema Registry failed")
```
Comment (3AceShowHand) [inline cdc/sink/codec/schema_registry.go:94]: ditto

File: cdc/sink/codec/schema_registry.go [hunk 27]
```diff
@@ -71,68 +69,102 @@ type lookupResponse struct {
...
+func (m *AvroSchemaManager) Register(
+	ctx context.Context,
+	topicName string,
+	codec *goavro.Codec,
+) (int, error) {
 	// The Schema Registry expects the JSON to be without newline characters
+	buffer := new(bytes.Buffer)
+	err := json.Compact(buffer, []byte(codec.Schema()))
+	if err != nil {
+		return 0, errors.Annotate(
```
Comment (3AceShowHand) [inline cdc/sink/codec/schema_registry.go:132]: ditto

File: cdc/sink/codec/schema_registry.go [hunk 28]
```diff
@@ -71,68 +69,102 @@ type lookupResponse struct {
...
+	}
 	reqBody := registerRequest{
-		Schema: regexRemoveSpaces.ReplaceAllString(codec.Schema(), ""),
-		// Commented out for compatibility with Confluent 5.4.x
-		// SchemaType: "AVRO",
+		Schema: buffer.String(),
 	}
 	payload, err := json.Marshal(&reqBody)
 	if err != nil {
 		return 0, errors.Annotate(
```
Comment (3AceShowHand) [inline cdc/sink/codec/schema_registry.go:145]: ditto

File: cdc/sink/codec/schema_registry.go [hunk 29]
```diff
@@ -144,24 +176,37 @@ func (m *AvroSchemaManager) Register(ctx context.Context, tableName model.TableN
...
+		return 0, cerror.ErrAvroSchemaAPIError.GenWithStack(
+			"Failed to register schema to the Registry, HTTP error",
+		)
 	}
 
 	var jsonResp registerResponse
 	err = json.Unmarshal(body, &jsonResp)
 
 	if err != nil {
 		return 0, errors.Annotate(
```
Comment (3AceShowHand) [inline cdc/sink/codec/schema_registry.go:196]: ditto

File: cdc/sink/codec/schema_registry.go [hunk 30]
```diff
@@ -213,7 +272,12 @@ func (m *AvroSchemaManager) Lookup(ctx context.Context, tableName model.TableNam
 	body, err := io.ReadAll(resp.Body)
 	if err != nil {
 		return nil, 0, errors.Annotate(
```
Comment (3AceShowHand) [inline cdc/sink/codec/schema_registry.go:274]: ditto

File: cdc/sink/codec/schema_registry.go [hunk 31]
```diff
@@ -222,22 +286,31 @@ func (m *AvroSchemaManager) Lookup(ctx context.Context, tableName model.TableNam
...
-		return nil, 0, cerror.ErrAvroSchemaAPIError.GenWithStackByArgs("Schema not found in Registry")
+		return nil, 0, cerror.ErrAvroSchemaAPIError.GenWithStackByArgs(
+			"Schema not found in Registry",
+		)
 	}
 
 	var jsonResp lookupResponse
 	err = json.Unmarshal(body, &jsonResp)
 	if err != nil {
 		return nil, 0, errors.Annotate(
```
Comment (3AceShowHand) [inline cdc/sink/codec/schema_registry.go:307]: ditto

File: cdc/sink/codec/config_test.go (hunk 24 above)
Comment (zhangyangyu) [reply cdc/sink/codec/config_test.go:40]: `NoError` is better. 1. It shows more semantics. 2. When fail, no error will show the trace of the error but `Nil` won't.

**Review by 3AceShowHand — APPROVED** [review]

File: pkg/cmd/cli/cli_changefeed_create.go [hunk 32]
```diff
@@ -136,6 +137,8 @@ func (o *createChangefeedOptions) addFlags(cmd *cobra.Command) {
 	cmd.PersistentFlags().BoolVarP(&o.disableGCSafePointCheck, "disable-gc-check", "", false, "Disable GC safe point check")
 	cmd.PersistentFlags().Uint64Var(&o.startTs, "start-ts", 0, "Start ts of changefeed")
 	cmd.PersistentFlags().StringVar(&o.timezone, "tz", "SYSTEM", "timezone used when checking sink uri (changefeed timezone is determined by cdc server)")
+	cmd.PersistentFlags().
+		StringVar(&o.schemaRegistry, "schema-registry", "", "Avro Schema Registry uri")
```
Comment (hi-rustin) [inline pkg/cmd/cli/cli_changefeed_create.go:141]:
```suggestion
		StringVar(&o.schemaRegistry, "schema-registry", "", "Avro Schema Registry URI")
```

**Review by hi-rustin — APPROVED** [review]
